web/environments.go

```diff
@@ -102,29 +106,140 @@ func NewLandscapeHandler(client consul.Client) gin.HandlerFunc {
 	}
 }
 
-func NewSAPSystemListHandler(client consul.Client) gin.HandlerFunc {
-	return func(c *gin.Context) {
-		var envName, landName string
+type SAPSystemRow struct {
+	SID            string
+	InstancesTable []*InstanceRow
+}
 
-		query := c.Request.URL.Query()
-		if len(query["environment"]) > 0 {
-			envName = query["environment"][0]
+type InstanceRow struct {
+	SID            string
+	InstanceNumber string
+	Features       string
+	Host           string
+	ClusterName    string
+	ClusterId      string
+}
+
+type SAPSystemsTable []*SAPSystemRow
+
+func NewSAPSystemsTable(sapSystemsList sapsystem.SAPSystemsList, hostList hosts.HostList) SAPSystemsTable {
+	var sapSystemsTable SAPSystemsTable
+	rowsBySID := make(map[string]*SAPSystemRow)
+
+	for _, s := range sapSystemsList {
+		if s.Type != sapsystem.Application {
+			continue
 		}
 
-		if len(query["landscape"]) > 0 {
-			landName = query["landscape"][0]
+		sapSystem, ok := rowsBySID[s.SID]
+		if !ok {
+			sapSystem = &SAPSystemRow{
+				SID: s.SID,
+			}
+			rowsBySID[s.SID] = sapSystem
 		}
 
-		environments, err := environments.Load(client)
+		for _, i := range s.Instances {
+			var features string
+			var instanceNumber string
+			var clusterName, clusterId string
+
+			if p, ok := i.SAPControl.Properties["SAPSYSTEM"]; ok {
+				instanceNumber = p.Value
+			}
+
+			for _, ci := range i.SAPControl.Instances {
+				if instanceNumber == fmt.Sprintf("%02d", ci.InstanceNr) {
+					features = ci.Features
+				}
+			}
+
+			for _, h := range hostList {
+				if i.Host == h.Name() {
+					clusterName = h.TrentoMeta()["trento-ha-cluster"]
+					clusterId = h.TrentoMeta()["trento-ha-cluster-id"]
+				}
+			}
+
+			instance := &InstanceRow{
+				SID:            s.SID,
+				InstanceNumber: instanceNumber,
+				Features:       features,
+				Host:           i.Host,
+				ClusterName:    clusterName,
+				ClusterId:      clusterId,
+			}
+
+			sapSystem.InstancesTable = append(sapSystem.InstancesTable, instance)
+		}
+	}
+
+	for _, row := range rowsBySID {
+		sapSystemsTable = append(sapSystemsTable, row)
+	}
+
+	return sapSystemsTable
+}
+
+func (t SAPSystemsTable) filter(sid []string) SAPSystemsTable {
```

Comment:
Maybe `filterBySid` would be a better name

Reply:
we will need to filter by health and possibly tag in the near future, i would leave it as `filter` also to match the `ClustersTable` filter function